File: nodejs-typescript/src/infrastructure/CsvAccountBalanceReader.ts
import { Account } from '../domain/Account.js';
import { AccountNumber } from '../domain/AccountNumber.js';
import { Money } from '../domain/Money.js';
import type { AccountBalanceReader } from '../application/LoadAccountBalances.js';
import { CsvParseError, csvParseError } from './errors.js';
import { columnIndex, readCsvRecords, rowNumber } from './csv.js';

export class CsvAccountBalanceReader implements AccountBalanceReader {
  constructor(private readonly path: string) {}

  eachAccount(callback: (account: Account) => void): void {
    for (const account of this.read()) {
      callback(account);
    }
  }

  read(): Account[] {
    let records: ReturnType<typeof readCsvRecords>;
    try {
      records = readCsvRecords(this.path);
    } catch (error) {
      throw csvParseError(`Account balances file not found: ${this.path}`, error);
    }

    const accountIndex = columnIndex(records.headers, 'Account');
    const balanceIndex = columnIndex(records.headers, 'Balance');
    return records.rows.map((row, index) =>
      this.parseRow(row, rowNumber(index), accountIndex, balanceIndex),
    );
  }

  private parseRow(
    row: string[],
    lineNumber: number,
    accountIndex: number,
    balanceIndex: number,
  ): Account {
    const accountValue = row[accountIndex];
    const balanceValue = row[balanceIndex];
    if (accountValue === undefined || balanceValue === undefined) {
      throw new CsvParseError(`Row ${String(lineNumber)}: missing Account or Balance`);
    }

    try {
      const accountNumber = AccountNumber.parse(accountValue);
      const balance = Money.fromDecimalString(balanceValue);
      return new Account(accountNumber, balance);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw csvParseError(`Row ${String(lineNumber)}: ${message}`, error);
    }
  }
}

File: ruby/lib/mable_bank/infrastructure/csv_account_balance_reader.rb
# frozen_string_literal: true

require 'csv'

module MableBank
  module Infrastructure
    class CsvAccountBalanceReader
      class ParseError < StandardError; end

      def initialize(path)
        @path = path
      end

      def each_account
        return enum_for(:each_account) unless block_given?

        CSV.foreach(@path, headers: true).with_index(2) do |row, row_number|
          yield parse_row(row, row_number)
        end
      rescue Errno::ENOENT
        raise ParseError, "Account balances file not found: #{@path}"
      rescue CSV::MalformedCSVError => e
        raise ParseError, "Malformed account balances CSV: #{e.message}"
      end

      def read
        each_account.to_a
      end

      private

      def parse_row(row, row_number)
        account_value = row['Account']
        balance_value = row['Balance']

        raise ParseError, "Row #{row_number}: missing Account or Balance" if account_value.nil? || balance_value.nil?

        account_number = Domain::AccountNumber.new(account_value)
        balance = Domain::Money.new(balance_value)

        Domain::Account.new(account_number: account_number, balance: balance)
      rescue Domain::InvalidAccountNumberError, Domain::InvalidMoneyAmountError => e
        raise ParseError, "Row #{row_number}: #{e.message}"
      end
    end
  end
end
